### ISSUE TITLE
🐛 compiler: unable to resolve commonjs named exports from an es module

### ISSUE THREAD

**Issue by dctalbot** [rome/tools#1402]
## Environment

```
Environment Variables
  {COLORFGBG: "7;0", COLORTERM: "truecolor", LANG: "en_US.UTF-8", TERM: "xterm-256color", TERM_PROGRAM: "vscode"}

User Config
  unset

Rome Version
  10.0.7-beta.18

Node Version
  14.16.0

Platform
  darwin x64 20.3.0

Terminal Features
  {
    background: "unknown"
    colorDepth: 24
    cursor: true
    hyperlinks: true
    isTTY: true
    progress: true
    unicode: true
    columns: OneIndexedNumber<153>
  }

Client Flags
  {
    clientName: "cli"
    silent: false
    cwd: AbsoluteFilePath</Users/dctalbot/Developer/wcbn/player-react-native>
    realCwd: AbsoluteFilePath</Users/dctalbot/Developer/wcbn/player-react-native>
  }

Server Status
  {
    projects: [{id: 0}, {id: 1}]
    workers: [{astCacheSize: 0, heapTotal: 32_333_824, ownedBytes: 0n, ownedFileCount: 0, pid: 55_220, uptime: 0.269131805}]
    server: {heapTotal: 32_333_824, pid: 55_220, uptime: 0.269205877}
  }

```

## Steps To Reproduce

This is an issue for probably any react-native project. Still working on creating a smaller example, but here is the scenario:

1. Have some commonjs resource that exposes exports like so:
```js
module.exports = {
  // ...
  get SafeAreaView(): SafeAreaView {
    return require('./Libraries/Components/SafeAreaView/SafeAreaView');
  },
  get Switch(): Switch {
    return require('./Libraries/Components/Switch/Switch');
  },
  get Dimensions(): Dimensions {
    return require('./Libraries/Utilities/Dimensions');
  },
  // ...
}
```

2. Try importing into an ES module
```js
import {Dimensions} from "react-native";
```

3. See lint error when importing one

```
src/styles/main.ts:1:8 resolver/unknownExport ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━

  ✖ Couldn't find export Dimensions in node_modules/react-native/index.js

  > 1 │ import {Dimensions} from "react-native";
      │         ^^^^^^^^^^
    2 │ 
    3 │ export const dimensions = {

  ℹ This file doesn't have any exports

```

## Expected Results
No error for this use case.

Note that the following would resolve the issue I think, but the onus probably shouldn't be on the developer to fix this everywhere. I don't know enough about all the different module formats to know what is best though.
```js
import * as ReactNative from "react-native";
const {Dimensions} = ReactNative;
```

**Comment by yuzima** [comment]
@dctalbot I think it's because `react-native` is using CommonJS which not support `module.exports` as named export.

**Comment by ljharb** [comment]
CommonJS supports named exports when transpiled with babel or typescript, and in modern node, which infers them vis cjs-module-lexer.

**Comment by yuzima** [comment]
@ljharb You are right, the named export is supported. I debugged locally noticed that react-native is using Flow, could this be the problem? As Rome doesn't support Flow now.

**Comment by ljharb** [comment]
node only deals with JavaScript, so any transpiler that produces known module named export patterns will be recognized, like Babel’s __esModule.

**Comment by jamiebuilds** [comment]
Found another issue when looking into this, the compiled output is not great:

```js
// test-package/index.js
module.exports = {
  foo: true,
};

// index.ts
import {foo} from "test-package"
foo;
```

```js
// uid://test-rome/test-package/index.js
var ___R$test$rome$test$package$index_js = {};
___R$test$rome$test$package$index_js = {
  foo: true,
};

// uid://test-rome/index.ts
const ___R$test$rome$index_ts = {};
___R$test$rome$test$package$index_js$foo; // <<<<
```

`...$index_js$foo` should be `...$index_js.foo` or something

**Comment by ktfth** [comment]
@jamiebuilds the output are coming from v8 error frames, that is my guess

internal/v8/error-frames.ts
113:    } else if (name.startsWith("___R$")) {

**Comment by ktfth** [comment]
I have changed some files to catch the point of this production

```
export function getPrefixedNamespace(moduleId: UIDPath): string {
	return `${SCOPE_PRIVATE_PREFIX}$debug$${normalizeModuleId(moduleId).format()}`;
}
```

```
(function(res) {
        if (typeof module !== "undefined") {
                module.exports = res;
        }
        return res;
})(
(function(global) {
  'use strict';
  // uid://test-rome/test-package/index.js
  var ___R$$debug$test$rome$test$package$index_js = {};
        ___R$$debug$test$rome$test$package$index_js = {
                foo: true,
        };


  // uid://test-rome/index.ts
        const ___R$$debug$test$rome$index_ts = {};
        ___R$$debug$test$rome$test$package$index_js$foo;


  return ___R$$debug$test$rome$index_ts;
})(typeof global !== 'undefined' ? global : typeof window !== 'undefined' ? window : this));
//# sourceMappingURL=index.ts.map
```

**Comment by ktfth** [comment]
internal\compiler\transforms\compileForBundle\modern\esToRefTransform.ts works on the compilation of the rome code
internal\compiler\transforms\compileForBundle\modern\cjsRootTransform.ts works on the bundle process

**Comment by dctalbot** [comment]
It seems like the compiler problem might be a separate issue. I don't think the linter would be transforming code, does that sound right?

**Comment by ematipico** [comment]
Exactly, this is a compiler problem. If you check the name of the diagnostic, it doesn't contain "lint" in its name.

**Comment by UsamaHameed** [comment]
I would love to take this on after I finish #1526

**Comment by jamiebuilds** [comment]
Hi! Thank you for opening this issue

We recently made the decision to [migrate the Rome codebase to Rust](https://rome.tools/blog/2021/09/21/rome-will-be-rewritten-in-rust) and making some significant changes to its architecture. We've also decided not to release the previous version written in TypeScript.

We're going through and closing pull requests and issues that were specific to the TypeScript codebase, including this one. Thank you for taking the time to open it.

Thanks again!
